flocker/control/test/test_generations.py

```diff
@@ -0,0 +1,82 @@
+# Copyright ClusterHQ Inc.  See LICENSE file for details.
+
+"""
+Tests for ``flocker.node._generations``.
+"""
+
+from testtools.matchers import Equals, Is, Not
+
+from ...testtools import TestCase
+
+from ..testtools import deployment_strategy
+from .._generations import GenerationTracker
+from .._persistence import make_generation_hash
+
+
+class GenerationTrackerTests(TestCase):
+
+    def test_basic_use_works(self):
+        """
+        After inserting a bunch of deployments into a ``GenerationTracker`` in
+        sequence, the ``Diff`` returned from ``get_diff_from_hash_to_latest``
+        can be applied to convert each of the deployments to the latest
+        deployment.
+        """
+        deployments = list(deployment_strategy().example() for _ in xrange(5))
+        deployments[3] = deployments[1]
```

Comment:
Yea... it was kind of convoluted. Added comments.

The code above is a bit more interesting if the queue has repeated deployments (the line labelled `if record.generation_hash == generation_hash:` gets hit multiple times).

Most of the other things are just trying to double check that inserting the same object as the latest a bunch is a no-op, and that you always get the same diff out of the tracker.